src/streaming/MediaPlayer.js

```diff
@@ -1288,6 +1290,79 @@ function MediaPlayer() {
+     * @memberof module:MediaPlayer
+     * @instance
+     */
+    function addTextTrack(url, mediaInfo) {
+        if (!streamingInitialized) {
+            throw STREAMING_NOT_INITIALIZED_ERROR;
+        }
+
+        if (textController === undefined) {
+            textController = TextController(context).getInstance();
+        }
+
+        /********* END ***********/
+        let captionsLoader = createCaptionsLoader();
+        let self = this;
+
+        const handler = function (e) {
+            if (!e.error) {
+                let textTracks = TextTracks(context).getInstance();
+
+                const textTrackInfo = new TextTrackInfo();
+                const trackKindMap = { subtitle: 'subtitles', caption: 'captions' }; //Dash Spec has no "s" on end of KIND but HTML needs plural.
+                const getKind = function () {
+                    let kind = (mediaInfo.roles.length > 0) ? trackKindMap[mediaInfo.roles[0]] : trackKindMap.caption;
+                    kind = (kind === trackKindMap.caption || kind === trackKindMap.subtitle) ? kind : trackKindMap.caption;
+                    return kind;
+                };
+
+                const checkTTML = function () {
+                    let ttml = false;
+                    if (mediaInfo.codec && mediaInfo.codec.search(Constants.STPP) >= 0) {
+                        ttml = true;
+                    }
+                    if (mediaInfo.mimeType && mediaInfo.mimeType.search(Constants.TTML) >= 0) {
+                        ttml = true;
+                    }
+                    return ttml;
+                };
+
+                textTrackInfo.captionData = e.captions;
+                textTrackInfo.lang = mediaInfo.lang;
+                textTrackInfo.labels = mediaInfo.labels;
+                textTrackInfo.id = mediaInfo.id ? mediaInfo.id : mediaInfo.index; // AdaptationSet id (an unsigned int) as it's optional parameter, use mediaInfo.index
+                textTrackInfo.index = mediaInfo.index; // AdaptationSet index in manifest
+                textTrackInfo.isTTML = checkTTML();
+                textTrackInfo.defaultTrack = false; // getIsDefault(mediaInfo);
+                textTrackInfo.isFragmented = false; // !adapter.getIsTextTrack(mediaInfo.mimeType);
+                textTrackInfo.isEmbedded = mediaInfo.isEmbedded ? true : false;
+                textTrackInfo.kind = getKind();
+                textTrackInfo.roles = mediaInfo.roles;
+                textTrackInfo.accessibility = mediaInfo.accessibility;
+                const totalNrTracks = textTracks.getNumberOfTextTracks() + 1;
+                textTracks.addTextTrack(textTrackInfo, totalNrTracks);
+            } else {
+                // ????
+            }
+            eventBus.off(Events.EXTERNAL_CAPTIONS_LOADED, handler, self);
+            captionsLoader.reset();
+        };
+
+        eventBus.on(Events.EXTERNAL_CAPTIONS_LOADED, handler, self);
```

Comment:
Can this event be captured by the TextSourceBuffer instead and cause createTextTrackFromMediaInfo() to be called? Instead of duplicating the code?

Reply:
You are totally right, I think the current structure is an artifact of how I wrote the change initially.

Change done in https://github.com/Dash-Industry-Forum/dash.js/pull/3573/commits/39773bba3849b7c361447b4d6229b641ae21f18d